pkg/controller/plugins.go

```diff
@@ -0,0 +1,144 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/errors"
+	"github.com/golang/glog"
+)
+
+// ControllerHost is an interface that plugins can use to access the kube-controller-manager
+type ControllerHost interface {
+	// GetKubeClient returns a client interface
+	GetKubeClient() client.Interface
+
+	// SyncPeriod asks the host for the sync period of the controller plugin with the given name.
+	// An error is returned if a period does not exist for the given name.
+	SyncPeriod(name string) (*time.Duration, error)
+
+	// SetSyncPeriod allows the plugin to set a default period with the host, also allows the host to override.
+	SetSyncPeriod(name string, period time.Duration)
+}
+
+// ControllerPluginManager tracks and manages the lifecycle of registered plugins
+type ControllerPluginManager struct {
+	mutex        sync.Mutex
+	plugins      map[string]ControllerPlugin
+	stopChannels map[string]chan struct{}
+}
+
+// InitPlugins initializes each plugin.  All plugins must have unique names.
+// This must be called exactly once before any New* methods are called on any plugins.
+func (mgr *ControllerPluginManager) InitPlugins(controllerPlugins []ControllerPlugin, host ControllerHost) error {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+
+	if mgr.plugins == nil {
+		mgr.plugins = map[string]ControllerPlugin{}
+		mgr.stopChannels = make(map[string]chan struct{})
+	}
+
+	allErrs := []error{}
+	for _, plugin := range controllerPlugins {
+		name := plugin.Name()
+		if !util.IsQualifiedName(name) {
+			allErrs = append(allErrs, fmt.Errorf("controller plugin has invalid name: %#v", plugin))
+			continue
+		}
+
+		if _, found := mgr.plugins[name]; found {
+			allErrs = append(allErrs, fmt.Errorf("controller plugin %q was registered more than once", name))
+			continue
+		}
+		plugin.Init(host)
+		host.SetSyncPeriod(plugin.Name(), plugin.DefaultSyncPeriod())
+		mgr.plugins[name] = plugin
+		glog.V(1).Infof("Loaded controller plugin %q", name)
+	}
+	return errors.NewAggregate(allErrs)
+}
+
+// FindPluginByName fetches a plugin by name.  If no plugin is found, returns error.
+func (mgr *ControllerPluginManager) FindPluginByName(name string) (ControllerPlugin, error) {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+
+	if plugin, ok := mgr.plugins[name]; ok {
+		return plugin, nil
+	}
+
+	return nil, fmt.Errorf("no controller plugin found matching %s", name)
+}
+
+// RunAll runs all Controllers in its own Go routine
```

Comment:
nit: "runs each Controller in its own goroutine" or "runs all Controllers in their own goroutines"

goroutine, not "Go routine"